File: core/src/main/java/com/yopox/ui/Text.java
package com.yopox.ui;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.yopox.Util;

import java.util.Arrays;
import java.util.List;
import java.util.Vector;
import java.util.stream.Collectors;

public class Text {
    public static final Texture mrmoTexture = new Texture("MRMOTEXT.png");
    public static TextureRegion[][] mrmoRegions = TextureRegion.split(mrmoTexture, Util.TILE.intValue(), Util.TILE.intValue());

    public static void draw(Vector<Letter> text, Float x, Float y, boolean yShift, SpriteBatch batch) {
        final List<Integer> lengths = Arrays.stream(Letter.stringOf(text).split("\n")).map(s -> s.length()).collect(Collectors.toList());
        if (lengths.isEmpty()) return;
        final Float maxWidth = lengths.stream().max((i1, i2) -> i2 - i1).get() * Util.TILE;
        final Vector<Float> xOffsets = lengths.stream()
                .map(i -> (maxWidth - i * Util.TILE) / 2)
                .collect(Collectors.toCollection(Vector::new));
        final Float xOffset = x == -1f ? (Util.WIDTH - maxWidth) / 2 : x;
        final Float yOffset = yShift ? Letter.heightOf(text) / 2 : 0f;

        drawStep(text, xOffsets, xOffset, y + yOffset, batch);
    }

    private static void drawStep(
            Vector<Letter> text,
            Vector<Float> xOffsets,
            Float x,
            Float y,
            SpriteBatch batch) {
        final Vector2 offset = new Vector2();
        for (final Letter l : text) {
            if (l.getCharacter() == '\n') {
                offset.x = 0f;
                offset.y -= 2 * Util.TILE;
                xOffsets.remove(0);
            } else {
                batch.draw(
                        l.getTexture(),
                        x + offset.x + xOffsets.get(0),
                        y + offset.y
                );
                offset.x += Util.TILE;
            }
        }
    }
}

File: core/src/main/kotlin/com/yopox/ui/Text.kt
package com.yopox.ui

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2
import com.yopox.TILE
import com.yopox.WIDTH
import ktx.math.plus
import ktx.math.vec2

val mrmoTexture = Texture("MRMOTEXT.png")
val mrmoRegions = TextureRegion.split(mrmoTexture, TILE.toInt(), TILE.toInt())

object Text {

    fun draw(text: List<Letter>, posY: Float, batch: SpriteBatch) {
        val offset = Vector2()
        val len = text.string().split("\n").map { it.length }
        val xOffsets = len.map { (len.maxOf { i -> i } - it) / 2 * TILE }.toMutableList()
        val centerOffset = (WIDTH - (len.maxOf { it } * TILE)) / 2
        val ySize = (2 * len.size - 1) * TILE

        drawStep(text, offset, xOffsets, vec2(centerOffset, posY + ySize / 2 - TILE), batch)
    }

    fun draw(text: List<Letter>, pos: Vector2, batch: SpriteBatch) {
        val offset = Vector2()
        val len = text.string().split("\n").map { it.length }
        val xOffsets = len.map { (len.maxOf { i -> i } - it) / 2 * TILE }.toMutableList()

        drawStep(text, offset, xOffsets, pos, batch)
    }

    private fun drawStep(
        text: List<Letter>,
        offset: Vector2,
        xOffsets: MutableList<Float>,
        pos: Vector2,
        batch: SpriteBatch
    ) {
        for (c in text) {
            if (c.char == '\n') {
                offset.x = 0f
                offset.y -= 2 * TILE
                xOffsets.removeAt(0)
            } else {
                draw(c, pos + offset + vec2(xOffsets.first(), 0f), batch)
                offset.x += TILE
            }
        }
    }

    fun draw(letter: Letter, pos: Vector2, batch: SpriteBatch) {
        val indexes = letter.indexes()
        batch.draw(mrmoRegions[indexes.first][indexes.second], pos.x, pos.y)
    }

}